lib/handler/traffic.c

```diff
@@ -0,0 +1,144 @@
+#include <stdlib.h>
+#include "h2o.h"
+
+#ifndef HUNDRED_MS
+#define HUNDRED_MS 100
+#endif
+
+#ifndef ONE_SECOND
+#define ONE_SECOND 1000
+#endif
+
+typedef H2O_VECTOR(h2o_iovec_t) iovec_vector_t;
+
+typedef struct st_traffic_shaper_t {
+    h2o_ostream_t super;
+    h2o_timeout_entry_t timeout_entry;
+    size_t tokens;
+    size_t token_inc;
+    h2o_context_t *ctx;
+    h2o_req_t *req;
+    struct {
+        iovec_vector_t bufs;
+        size_t inbufcnt;
+        int is_final;
+    } state;
+} traffic_shaper_t;
+
+static int real_send(traffic_shaper_t *self) {
+    /* a really simple token bucket implementation */
+    assert(self->tokens);
+    size_t i, token_consume;
+
+    token_consume = 0;
+
+    for (i = 0; i < self->state.inbufcnt; i++) {
+        if (self->state.bufs.entries[i].len > self->tokens - token_consume) {
+            return 0;
+        }
+        token_consume += self->state.bufs.entries[i].len;
+    }
+
+    self->tokens -= token_consume;
+
+    h2o_ostream_send_next(&self->super, self->req, self->state.bufs.entries,
+                          self->state.inbufcnt, self->state.is_final);
+    return self->state.is_final;
+}
+
+static inline void expand_buf(h2o_mem_pool_t *pool, iovec_vector_t *bufs, size_t count) {
+    assert(bufs->size < count);
+    h2o_vector_reserve(pool, bufs, count);
+    bufs->size = count;
+}
+
+static void add_token(h2o_timeout_entry_t *entry) {
+    traffic_shaper_t *shaper = H2O_STRUCT_FROM_MEMBER(traffic_shaper_t, timeout_entry, entry);
+    shaper->tokens += shaper->token_inc;
+
+    if (!real_send(shaper))
+        h2o_timeout_link(shaper->ctx->loop, &shaper->ctx->hundred_ms_timeout, &shaper->timeout_entry);
+}
+
+static void on_send(h2o_ostream_t *_self, h2o_req_t *req, h2o_iovec_t *inbufs, size_t inbufcnt, int is_final) {
+    traffic_shaper_t *self = (void *)_self;
+    size_t i;
+
+    /* I don't know if this is a proper way. */
+    if (!h2o_timeout_is_linked(&self->timeout_entry)) {
+        h2o_timeout_link(self->ctx->loop, &self->ctx->hundred_ms_timeout, &self->timeout_entry);
+    }
```

Comment:
`add_token` should be called at constant pace while the filter is open.  In other words, I believe you should:

* call `h2o_timeout_link` in `on_setup_stream` (when the filter is opened), and in `add_token` (when the timeout gets fired)
* call `h2o_timeout_unlink` in `real_send` if `on_send` has received `is_final`
* call `h2o_timeout_unlink` (if timeout is linked) in `on_stop`